frontend: overhaul validator dashboard fetching logic

LuccaBitfly opened the pull request:
In the current validator dashboard fetching flow, components will either fetch their own data by reacting to changes in other unrelated components or manually trigger refreshes in unrelated components, making tracing bugs a hassle.
This PR aims to bring back order by making sure that most data fetching happens in the dashboards root component and that these are only triggered through emits. As a consequence of removing sibling component coupling, most of the dashboards Pinia stores won't be necessary anymore. 

### Edit 06/02/2025:
**Branch updated with following adjustments:**
- Add `resetOverviewData` function to `validatorDashboardStore` - needed when there's no dashboard key (guest dashboard with no validators), so `overview` isn't refetched and the store needs to reflect empty state
- Prevent expanded summary rewards rows from collapsing on `slot-viz` refetch - now data won't update on that table when polling `slot-viz` data, but this is the way it worked before this PR anyway - fixing this is a bigger issue involving changes in `BcTable` and all it's implementations by passing completely unique identifier to each row.
- Rename `isAllSelected` to `hasNoGroupsOrAllSelected` in dashboard root component
- Organize all logic using `inline composables` inside root component

LuccaBitfly committed:
refactor: overhaul overview and slotviz data flow

- Overview and SlotViz data is now fetched in `index.vue` and propped down
- SlotViz updates are handled through update emits
- Introduces proper SSR fetching
- removes all direct sibling dependencies on the overview (including the dashboard group composable)
- introduces new validator dashboard store
- dashboard modyfing actions now emit an `updateAll` event instead of directly calling `refreshOverview`

See: BEDS-914

marcel-bitfly commented on frontend/components/dashboard/DashboardValidatorOverview.vue:
```diff
@@ -1,18 +1,21 @@
 
-const { overview } = useValidatorDashboardOverviewStore()
+const props = defineProps<{
+  overviewData?: VDBOverviewData,
```
np: you could think about naming this `data` as you are in the `overview` component. That would make things a little less `verbose`. But `overviewData` is also fine.

LuccaBitfly replied:
will do

marcel-bitfly commented on frontend/composables/useValidatorDashboardOverview.ts:
np:
🤔 if this `composable` would ever only be used inside one component this might be the wrong abstraction or this might not be pulled out of the `component`

LuccaBitfly replied:
I think it makes sense to at least put the fetching logic outside of the component (e.g. a service file), as to not clutter the index component with too much code. I.e. inside the component, the only concern should be what you fetch and not how you fetch it. As we have no pattern for services right now, I'm opting to extract them into composables. Is that an antipattern?

marcel-bitfly replied:
not an anti-pattern per se, but what you tell the world is: 
- here is a `composable` plz feel free to implement me wherever you want
- I would come in handy in other `contexts` as well

marcel-bitfly commented on frontend/composables/useValidatorSlotViz.ts:
```diff
@@ -0,0 +1,44 @@
+export function useValidatorSlotViz() {
+  const { fetch } = useCustomFetch()
+
+  // State
+  const slotVizData = ref<SlotVizEpoch[]>() // Renamed `data` to `slotVizData` for clarity
+
+  // Actions
```
only add `comments` if they are really necessary

LuccaBitfly replied:
was a chatGPT residue, will remove

marcel-bitfly commented on frontend/composables/useValidatorSlotViz.ts:
```diff
@@ -0,0 +1,44 @@
+        headers: {},
+        query,
+      },
+      { dashboardKey: dashboardKey || 'MQ' }, // If guest dashboard has no validators yet (= empty dashboardKey), load small guest dashboard with 1 validator (MQ)
```
this one is nice 👌
As it gives more context and explains the `why`

marcel-bitfly replied:
is this what should happen though? (loading validator index 1 if empty?)

LuccaBitfly replied:
Yes, the goal is to load a very small slot viz and remove the validator information. This is hacky, but the only way to achieve an empty guest dashboard slot viz currently.

marcel-bitfly commented on frontend/pages/dashboard/[[id]]/index.vue:
```diff
@@ -175,6 +144,78 @@ watch(
   { immediate: true },
 )
+
+// ------------------- Overview -------------------
```
I get the idea of these comments, but in my experience they do not pay off.

Comment should never contain the `what` but only the `why`

LuccaBitfly replied:
will remove

marcel-bitfly commented on frontend/pages/dashboard/[[id]]/index.vue:
```diff
@@ -175,6 +144,78 @@ watch(
+
+// ------------------- Overview -------------------
+
+const {
+  fetchOverviewData,
+  overviewData,
+} = useValidatorDashboardOverview()
+
+// ------------------- Slot Viz -------------------
+const {
+  fetchSlotVizData,
+  slotVizData,
+} = useValidatorSlotViz()
+async function refreshSlotViz(groupIds?: number[]) {
+  const data = await fetchSlotVizData(dashboardKey.value, groupIds)
+  slotVizData.value = data
+}
```
if there is a pattern here:
- fetching `endpoint`
- serving `data`
there might be a better abstraction here 🤔

LuccaBitfly replied:
if you have anything specific in mind, please lmk

marcel-bitfly commented on frontend/pages/dashboard/[[id]]/index.vue:
```diff
@@ -175,6 +144,78 @@ watch(
+// Helper: Handle errors
+function handleError(e: any) {
+  if (e?.statusCode === 404) {
+    // TODO: redirect to 404 page
```
why `404` page exists?

LuccaBitfly replied:
not immediately relevant anymore

marcel-bitfly commented on frontend/pages/dashboard/[[id]]/index.vue:
```diff
@@ -175,6 +144,78 @@ watch(
+// Helper: fetches all data for the dashboard (overview, slot viz, active table)
+function fetchAllData() {
+  if (dashboardKey.value) { // valid dashboard key -> fetch all data
+    return Promise.all([
```
plz think about `Promise.allSettled` instead: https://dev.to/viclafouch/promise-allsettled-vs-promise-all-in-javascript-4mle

proper error handling would be nice (but I guess is still wip)

LuccaBitfly replied:
implemented

LuccaBitfly replied:
changed to use `useAsyncData`

marcel-bitfly commented on frontend/components/dashboard/DashboardControls.vue:
```diff
@@ -43,11 +42,12 @@ const {
-  dashboards, getDashboardLabel, refreshDashboards, updateGuestDashboardKey,
-}
-  = useUserDashboardStore()
+  getDashboardLabel, refreshDashboards, updateGuestDashboardKey,
+} = userDashboardStore
+
+const { dashboards } = storeToRefs(userDashboardStore)
```
🤔 still don't know if I like this pattern more than:

const store = useUserDashboardStore()
//...
store.getDashboardLabel()
//...
store.dashboards

there is a lot going on in line `45` to `50` isn't there?

LuccaBitfly replied:
I have no strong opinion on this. This pattern is given in the pinia doc https://pinia.vuejs.org/core-concepts/#Destructuring-from-a-Store but I'm fine with accessing everything through the store const.

marcel-bitfly replied:
I trust your `gut feeling`...as both work fine

marcel-bitfly commented on frontend/components/dashboard/GroupManagementModal.vue:
```diff
@@ -116,7 +116,7 @@ const addGroup = async () => {
     { dashboardKey: dashboardKey.value },
   )
-  await refreshOverview(dashboardKey.value)
+  emit('updateAll')
```
why is the `DashboardGroupManagementModal` component `emitting` a `update-all` event?
How does this `component` know about things like `all`?

LuccaBitfly replied:
Something like `modified-dashboard` would prob be better naming, will change.

marcel-bitfly commented on frontend/stores/dashboard/useValidatorDashboardStore.ts:
```diff
@@ -0,0 +1,58 @@
+  'validator_dashboard_store',
+  () => {
+    const { t: $t } = useTranslation()
+    // TODO: bring ID in here and remove the provider composable!!
```
❤️ love this -> #promotioncantidate

enzo-bitfly replied:
That's where it belongs and we'll add it here when we remove the provider in the near future

marcel-bitfly reviewed (commented):
very nice work so far 💪

LuccaBitfly committed:
refactor: move dashboard table fetching to index

- Changes all dashboard table logic to emit update events instead of fetching data themselves
- Data is now fetched in the dashboard index component and propped down
- Removes all table stores
- Introduces dashboard composable to hold fetching logic
- Dashboard modifications now trigger only one table request

See: BEDS-914

marcel-bitfly commented on frontend/components/dashboard/DashboardControls.vue:
```diff
@@ -58,6 +57,13 @@ const isMobile = computed(() => width.value < 520)
+const emit = defineEmits<{
+  (e: 'modified-dashboard'): void,
+}>()
+function emitModifiedDashboard() {
+  emit('modified-dashboard')
+}
```
I would suggest `emitDashboardModified()`

See: https://github.com/kettanaito/naming-cheatsheet?tab=readme-ov-file#ahclc-pattern

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/DashboardValidatorManagementModal.vue:
```diff
@@ -270,11 +273,8 @@ const removeRow = (row: VDBManageValidatorsTableRow) => {
-  return addUpValues(overview.value?.validators as unknown as Record<string, number>)
-})
+
+const totalValidators = computed(() => validatorCount.value ?? 0)
```
maybe `validatorCount` should always be `0` -> default from `store`

enzo-bitfly replied:
I normally initialize values from fetched data as `null` to make clear that the data is not yet in the store and avoid hidden bugs.

marcel-bitfly commented on frontend/components/dashboard/DashboardValidatorOverview.vue:
```diff
@@ -1,18 +1,21 @@
 const { converter } = useValue()
 
-const { overview } = useValidatorDashboardOverviewStore()
+const props = defineProps<{
```
could be spread out now directly due to the newest update of vue 🥳

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/DashboardValidatorOverview.vue:
```diff
@@ -1,18 +1,21 @@
+const props = defineProps<{
+  data?: VDBOverviewData,
+}>()
+const { data } = toRefs(props)
```
this should be spreaded out, so that this component will not have just one prop where the details are hidden

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/ValidatorSlotViz.vue:
```diff
@@ -1,45 +1,46 @@
-    dashboardKey.value,
-    selectedGroups.value,
-    tick.value,
+  [
+    selectedGroups,
+    tick,
   ],
```
could these split into two different `watch`?

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/ValidatorSlotViz.vue:
```diff
@@ -1,45 +1,46 @@
 const {
-  resetTick, tick,
+  tick,
 } = useInterval(12)
```
this might not make sense in this component

LuccaBitfly replied:
moved outside

marcel-bitfly commented on frontend/components/dashboard/ValidatorSlotViz.vue:
```diff
@@ -50,16 +51,20 @@ const initiallyHideVisible = computed(() => {
   return orderBy(
-    overview.value.groups.filter(g => !!g.count),
+    dashboardGroups.value.filter(g => !!g.count),
     [ g => g.name.toLowerCase() ],
```
why is this even necessary?

enzo-bitfly replied:
It doesn't hurt to keep an alphabetical order, since this will make it easier to find groups if there are too many.

marcel-bitfly commented on frontend/components/dashboard/ValidatorSlotViz.vue:
```diff
@@ -50,16 +51,20 @@ const initiallyHideVisible = computed(() => {
   )
 })
 
+const isAllSelected = (groupList: number[]) => {
```
if this function is executed above the definition should be above the line

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/ValidatorSlotViz.vue:
```diff
@@ -1,45 +1,46 @@
-  refreshSlotViz, slotViz,
-} = useValidatorSlotVizStore()
+const props = defineProps<{
+  data?: SlotVizEpoch[],
```
should this even be **optional**?

LuccaBitfly replied:
no, changed

marcel-bitfly commented on frontend/components/dashboard/chart/DashboardChartSummaryFilter.vue:
```diff
@@ -13,10 +13,11 @@ import { getGroupLabel } from '~/utils/dashboard/group'
+const validatorDashboardStore = useValidatorDashboardStore()
 const {
-  hasAbilityCharthistory,
-  overview,
```
this could also be part of line `16`, as `validatorDashboarStore.hasAbilityChartHistory` would work too

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableBlocks.vue:
```diff
@@ -60,10 +58,7 @@ const loadData = (query?: TableQueryParams) => {
-    dashboardKey,
-    overview,
-  ],
+  [ dashboardKey ],
```
do you need this array?

LuccaBitfly replied:
no, changed it

marcel-bitfly commented on frontend/components/bc/tab/BcTabList.vue:
```diff
@@ -12,6 +12,16 @@ const props = defineProps<Props>()
+  (value) => {
+    emit('changed-tab', value)
+  },
+  { immediate: true },
```
this might be unnecessary 🤔

LuccaBitfly replied:
it is, changed

marcel-bitfly commented on frontend/components/bc/table/BcTable.vue:
```diff
@@ -1,11 +1,14 @@
   addSpacer?: boolean,
   cursor?: Cursor,
-  data?: ApiPagingResponse<any>,
+  data?: {
+    data?: any[],
+    paging?: Paging,
+  },
```
would be nice if this would be top level

marcel-bitfly replied:
this could also be a `generic component` using `<script lang="ts" generic="T">`

LuccaBitfly replied:
true, but IMO out of scope for this PR

enzo-bitfly replied:
I flattened these props, although the nesting wasn't introduced in this PR.
About the `generic component` - that's a great idea but should be done in a separate PR as this one is already too big.

marcel-bitfly commented on frontend/components/bc/table/TablePager.vue:
```diff
@@ -173,7 +173,7 @@ watch(
         </div>
       </template>
-      <Dropdown
+      <Select
```
thx this is a nice improvement 💪

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableBlocks.vue:
```diff
@@ -2,37 +2,32 @@
-  temp: tempQuery,
-  value: query,
-} = useDebounceValue<TableQueryParams | undefined>(undefined, 500)
+const props = defineProps<TableProps<VDBBlocksTableRow>>()
```
your hiding implementation details that are essential to work with this component  😭

LuccaBitfly replied:
changed

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableBlocks.vue:
```diff
@@ -2,37 +2,32 @@
-} = useDebounceValue<TableQueryParams | undefined>(undefined, 500)
+const props = defineProps<TableProps<VDBBlocksTableRow>>()
+const emit = defineEmits<{
+  (e: 'update', query: TableQueryParams): void,
```
would be nice to have a parent component for that

LuccaBitfly replied:
true, IMO out of scope for this PR as this will touch all tables including notifications, not just dashboards

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableBlocks.vue:
```diff
@@ -49,57 +44,26 @@ const colsVisible = computed(() => {
 
 const onSort = (sort: DataTableSortEvent) => {
-  loadData(setQuerySort(sort, lastQuery.value))
+  emitUpdate(setQuerySort(sort, props.query))
```
plz don't stop half way...find a better name for `setQuerySort....` as I do not understand what happens here

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableClDeposits.vue:
```diff
@@ -134,9 +107,8 @@ const isRowExpandable = (row: VDBConsensusDepositsTableRow) => {
             :row-class="getRowClass"
             :is-row-expandable
-            :loading="isLoadingDeposits"
+            :loading="isLoading"
```
```suggestion
            :is-loading
```

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableRewards.vue:
```diff
@@ -208,13 +148,8 @@ const findNextEpochDuties = (epoch: number) => {
-                </span>
                 <DashboardTableValueDuty
-                  v-else
+                  :is-number-visible="slotProps.data.group_id !== DAHSHBOARDS_NEXT_EPOCH_ID"
```
np: but `number` is also very `generic` here
```suggestion
                  :has-number="slotProps.data.group_id !== DAHSHBOARDS_NEXT_EPOCH_ID"
```

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableSummary.vue:
```diff
@@ -30,33 +32,25 @@ const chartFilter = ref<SummaryChartFilter>({
+  (e: 'update', timeframe: SummaryTimeFrame, query: TableQueryParams): void,
+}>()
+const emitUpdate = (query: TableQueryParams) => {
+  emit('update', props.timeFrame, query)
```
why is `timeframe` not part of the `query`

LuccaBitfly replied:
currently the summary table is the only component which uses timeframe, in future more components will use it. would suggest doing this when adding it to the other components so that this PR doesn't get more bloated

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableSummary.vue:
```diff
@@ -167,12 +123,13 @@ const searchPlaceholder = computed(() =>
           option-label="name"
           class="small"
           :placeholder="$t('dashboard.group.selection.placeholder')"
+          @select="($event) => { emit('update', $event, props.query) }"
```
you did tightly couple us here to an implementation detail of `prime vue` here....🤔

marcel-bitfly replied:
this might already work with `@update`

enzo-bitfly replied:
The event name wasn't changed in this MR and it belongs to `BcDropdown` not to `Select` from `vue prime`.
Still this line was changed and now looks like this: `@select="value => timeFrame = value"`

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableValueDuty.vue:
```diff
@@ -3,8 +3,14 @@ import type { VDBRewardsTableDuty } from '~/types/api/validator_dashboard'
+  isNumberVisible?: boolean,
 }
-defineProps<Props>()
+withDefaults(
+  defineProps<Props>(),
+  {
+    isNumberVisible: true,
+  },
+)
```
could be spread out natively now due to an `vue update` 💪

LuccaBitfly replied:
done

marcel-bitfly commented on frontend/components/dashboard/table/DashboardTableWithdrawals.vue:
```diff
@@ -156,6 +113,28 @@ const isRowInFuture = (row: ExtendedVDBWithdrawalsTableRow) => {
+      },
+      ...rows.map(w => ({
+        ...w,
+        identifier: `${w.slot}-${w.index}`,
```
could you use your generic `identifier function` you added before

enzo-bitfly replied:
Done

marcel-bitfly commented on frontend/composables/useDataFetcher.ts:
we talked briefly about this pattern, maybe it is best to stick to `useAsyncData` whenever possible...apart from that we should make `useFetch` work, or implement a `BFF`

LuccaBitfly replied:
done

LuccaBitfly committed:
refactor: rename query funcs

LuccaBitfly committed:
refactor: rename `loading` bool prop to `isLoading`

LuccaBitfly committed:
refactor: spread props in dashboard overview component

enzo-bitfly committed:
feat: keep slot-viz visible during loading state

- (BcLoadingSpinner) add background color to backdrop
- (DashboardSlotVizDutyVisibilityToggle) prevent hydration mismatch by using `fallback` slot in `ClientOnly` component
- Remove previous guest dashboard key from storage when the dashboard is updated

enzo-bitfly committed:
refactor: flatten `data` and `paging` props in `BcTable`

enzo-bitfly commented:
@marcel-bitfly please see `edit` in the PR description and go through open threads.
It may be helpful to take a fresh look at the `root` dashboard component

enzo-bitfly committed:
WIP: refactor dashboard index with inline composables
